langchain/src/tools/base.ts

```diff
@@ -1,39 +1,39 @@
-import { CallbackManager, getCallbackManager } from "../callbacks/index.js";
-
-const getVerbosity = () => false;
+import { CallbackManager } from "../callbacks/base.js";
+import { BaseLangChain } from "../base_language/index.js";
 
 export interface ToolParams {
   verbose?: boolean;
   callbackManager?: CallbackManager;
 }
 
-export abstract class Tool {
-  verbose: boolean;
-
-  callbackManager: CallbackManager;
-
+export abstract class Tool extends BaseLangChain {
   constructor(verbose?: boolean, callbackManager?: CallbackManager) {
-    this.verbose = verbose ?? (callbackManager ? true : getVerbosity());
-    this.callbackManager = callbackManager ?? getCallbackManager();
+    super(verbose, callbackManager);
   }
 
-  protected abstract _call(arg: string): Promise<string>;
-
-  async call(arg: string, verbose?: boolean): Promise<string> {
-    const _verbose = verbose ?? this.verbose;
-    await this.callbackManager.handleToolStart(
-      { name: this.name },
-      arg,
-      _verbose
-    );
+  protected abstract _call(
+    arg: string,
+    callbackManager?: CallbackManager
+  ): Promise<string>;
+
+  async call(
+    arg: string,
+    _verbose?: boolean,
+    callbackManager?: CallbackManager
+  ): Promise<string> {
+    const localCallbackManager = this.configureCallbackManager(callbackManager);
+    await localCallbackManager?.handleToolStart({ name: this.name }, arg);
+    if (callbackManager) {
+      callbackManager.setCurrentRunId(localCallbackManager?.currentRunId);
```

Comment:
Why?

Reply:
I agree it's a bit cumbersome but this is something that needs to just be done in the base classes I'm pretty sure. I don't envision users needing to do this in custom chains. The reason for doing this is to update the callbackmanager that has been passed through with the currentRunId of the callbackManager copy created with `configureCallbackManager`. We need to preserve the passed-through one to avoid leaking callback handlers in a parent class not present in a child class (like if llmchain has consolecallback but not llm)